neqo-qpack/src/encoder.rs

```diff
@@ -264,28 +278,56 @@ impl QPackEncoder {
         }
     }
 
-    fn change_capacity(&mut self, value: u64) -> Res<()> {
+    fn change_capacity(&mut self, value: u64) {
         qdebug!([self], "change capacity: {}", value);
-        self.table.set_capacity(value)?;
-        EncoderInstruction::Capacity { value }.marshal(&mut self.send_buf, self.use_huffman);
+        self.next_capacity = Some(value);
+    }
+
+    fn maybe_send_change_capacity(&mut self, conn: &mut Connection, stream_id: u64) -> Res<()> {
+        if let Some(cap) = self.next_capacity {
+            // Check if itt is possible to reduce the capacity, e.g. if enough space can be make free for the reduction.
+            if cap < self.table.capacity() && !self.table.test_evict_to(cap) {
+                return Err(Error::DynamicTableFull);
+            }
+            let mut buf = QPData::default();
+            EncoderInstruction::Capacity { value: cap }.marshal(&mut buf, self.use_huffman);
+            if !conn.stream_send_atomic(stream_id, &buf)? {
+                return Err(Error::EncoderStreamBlocked);
+            }
+            if self.table.set_capacity(cap).is_err() {
+                debug_assert!(
+                    false,
+                    "test_evict_to should have checked and make sure this operation is possible"
```

Comment:
If you change the name:
```suggestion
                    "can_evict_to should have checked and make sure this operation is possible"
```